skbio/sequence/_substitution.py

```diff
@@ -0,0 +1,518 @@
+    -1.0
+    >>> mat['G', 'G']
+    2.0
+    >>> blosum62 = SubstitutionMatrix.by_name('BLOSUM62')
+
+    """
+
+    @property
+    @experimental(as_of='0.5.10')
+    def alphabet(self):
+        """Alphabet of the substitution matrix.
+
+        Each element (character) corresponds to one row/column in the matrix.
+
+        Returns
+        -------
+        tuple
+            Alphabet of the substitution matrix.
+
+        Notes
+        -----
+        This is an alias of ``ids``.
+
+        """
+        return self._ids
+
+    @property
+    @experimental(as_of='0.5.10')
+    def scores(self):
+        """Matrix of substitution scores.
+
+        Each value corresponds to the score of substituting the row character
+        with the column character.
+
+        Returns
+        -------
+        2D np.ndarray
+            Matrix of substitution scores.
+
+        Notes
+        -----
+        This is an alias of ``data``.
+
+        """
+        return self._data
+
+    @experimental(as_of='0.5.10')
+    def __init__(self, alphabet, scores, **kwargs):
+        """Initialize a substitution matrix object with an alphabet and a score
+        matrix.
+        """
+        super().__init__(scores, alphabet, **kwargs)
+
+    @experimental(as_of='0.5.10')
+    def to_dict(self):
+        """Create a 2D dictionary from the substitution matrix.
+
+        Returns
+        -------
+        dict of dict
+            2D dictionary constructed from the substitution matrix.
+
+        """
+        return {id_: dict(zip(self._ids, row)) for id_, row in zip(
+            self._ids, self._data)}
+
+    @classonlymethod
+    @experimental(as_of='0.5.10')
+    def from_dict(cls, dictionary):
+        """Create a substitution matrix from a 2D dictionary.
+
+        Parameters
+        ----------
+        dictionary : dict of dict
+            2D dictionary of substitution scores from outer characters to inner
+            characters.
+
+        Returns
+        -------
+        SubstitutionMatrix
+            Substitution matrix constructed from the dictionary.
+
+        Raises
+        ------
+        ValueError
+            If outer and inner characters are inconsistent.
+        ValueError
+            If scores are not numbers.
+
+        Examples
+        --------
+        >>> from skbio import SubstitutionMatrix
+        >>> d = {'a': {'a': 1, 'b': 0, 'c': 0},
+        ...      'b': {'a': 0, 'b': 1, 'c': 0},
+        ...      'c': {'a': 0, 'b': 0, 'c': 1}}
+        >>> mat = SubstitutionMatrix.from_dict(d)
+        >>> mat.alphabet
+        ('a', 'b', 'c')
+        >>> mat.scores
+        array([[ 1.,  0.,  0.],
+               [ 0.,  1.,  0.],
+               [ 0.,  0.,  1.]])
+
+        """
+        alphabet = tuple(dictionary.keys())
+        alphabet_set = set(alphabet)
+        idmap = {x: i for i, x in enumerate(alphabet)}
+        scores = np.zeros((n := len(alphabet), n))
+        for i, row in enumerate(dictionary.values()):
```

Comment:
This implicitly assumes Python will ensure the order of evaluation of `.values` and `.keys` is stable and consistent. While it's true, this may be unexpected for developers unfamiliar with Python (or for those familiar only with old versions of Python). It may be pleasant to be explicit, use `.items`, and pull index directly from the `idmap` on line 196

Reply:
This is a good point. Let me modify.